tests/test_autoingest.py

```diff
@@ -0,0 +1,270 @@
+        name = self.strip_prefix(self.destFile, self.forwarderDir)
+        file_to_ingest = os.path.join(self.stagingDir, name)
+        self.assertTrue(os.path.exists(file_to_ingest))
+
+        # this file should now not exist
+        self.assertFalse(os.path.exists(self.destFile))
+
+        # add one more task, whose sole purpose is to interrupt the others by
+        # throwing an acception
+        task_list.append(asyncio.create_task(self.interrupt_me()))
+
+        # kick off all the tasks, until one (the "interrupt_me" task)
+        # throws an exception
+        try:
+            await asyncio.gather(*task_list)
+        except Exception:
+            for task in task_list:
+                task.cancel()
+
+        # that should have been enough time to run the "real" tasks,
+        # which performed the ingestion, and the clean up task, which
+        # was set to clean it up right away.  (That "clean up" time
+        # is set in the config file loaded for this FileIngester).
+        # And, when "cleaned up", the file that was originally there
+        # is now gone.  Check for that.
+        self.assertFalse(os.path.exists(file_to_ingest))
+
+        # check to be sure that the file wasn't "bad" (and
+        # therefore, not ingested)
+        bad_path = os.path.join(self.badDir, fits_name)
+        self.assertFalse(os.path.exists(bad_path))
+
+    async def testBadIngest(self):
+        """test ingesting a bad file
+        """
+        fits_name = "bad.fits.fz"
+        config = self.createConfig("ingest_comcam_gen3.yaml", fits_name)
+
+        # setup directory to scan for files in the forwarder staging directory
+        ingesterConfig = config["ingester"]
+        forwarder_staging_dir = ingesterConfig["forwarderStagingDirectory"]
+        scanner = DirectoryScanner([forwarder_staging_dir])
+        files = scanner.getAllFiles()
+        self.assertEqual(len(files), 1)
+
+        ingester = FileIngester(config["ingester"])
+
+        await ingester.ingest([self.destFile])
+
+        files = scanner.getAllFiles()
+        self.assertEqual(len(files), 0)
+
+        name = self.strip_prefix(self.destFile, forwarder_staging_dir)
+        bad_path = os.path.join(self.badDir, name)
+        self.assertTrue(os.path.exists(bad_path))
+
+    async def testRepoExists(self):
+        """test that a repository exists
+        """
+        fits_name = "bad.fits.fz"
+        config = self.createConfig("ingest_comcam_gen3.yaml", fits_name)
+
+        FileIngester(config["ingester"])
+        # tests the path that the previously created repo (above) exists
+        FileIngester(config["ingester"])
+
+    async def interrupt_me(self):
+        """Used to interrupt asyncio.gather() so that test can be halted
+        """
+        await asyncio.sleep(20)
+        raise RuntimeError("I'm interrupting")
+
+
+class MemoryTester(lsst.utils.tests.MemoryTestCase):
+    pass
+
+
+def setup_module(module):
+    lsst.utils.tests.init()
```

Comment:
Putting aside few extra comments or changed variable names, this test module is almost identical with `test_gen3.py`. Is there a some deeper reason behind having two of them?

Reply:
Yes, test_autoingest.py uses the asyncio mechanisms to do the ingest (and cleaning), where test_gen3.py does the same tests without using those mechanisms.